File: androidaudioplugin/src/main/cpp/core/hosting/gui-helper.cpp
#include "aap/core/host/gui-helper.h"
#include <algorithm>

namespace aap {

GuiViewportConfiguration GuiViewportConfiguration::clamp() const {
    GuiViewportConfiguration clamped = *this;
    clamped.viewportWidth = std::max<int32_t>(0, clamped.viewportWidth);
    clamped.viewportHeight = std::max<int32_t>(0, clamped.viewportHeight);
    clamped.contentWidth = std::max(clamped.viewportWidth, clamped.contentWidth);
    clamped.contentHeight = std::max(clamped.viewportHeight, clamped.contentHeight);
    auto maxScrollX = std::max<int32_t>(0, clamped.contentWidth - clamped.viewportWidth);
    auto maxScrollY = std::max<int32_t>(0, clamped.contentHeight - clamped.viewportHeight);
    clamped.scrollX = std::clamp(clamped.scrollX, 0, maxScrollX);
    clamped.scrollY = std::clamp(clamped.scrollY, 0, maxScrollY);
    return clamped;
}

GuiSize getGuiPreferredSizeOrFallback(
    RemotePluginInstance* instance,
    int32_t fallbackWidth,
    int32_t fallbackHeight) {
    GuiSize size{fallbackWidth, fallbackHeight};
    if (!instance)
        return size;
    instance->prepareSurfaceControlForRemoteNativeUI();
    int32_t preferredWidth = 0;
    int32_t preferredHeight = 0;
    if (instance->getRemoteNativeViewPreferredSize(preferredWidth, preferredHeight)) {
        if (preferredWidth > 0)
            size.width = preferredWidth;
        if (preferredHeight > 0)
            size.height = preferredHeight;
    }
    return size;
}

}

File: androidaudioplugin/src/main/java/org/androidaudioplugin/hosting/GuiHelper.kt
package org.androidaudioplugin.hosting

import android.content.Context
import android.os.Build
import android.view.View
import androidx.annotation.RequiresApi

object GuiHelper {
    data class Size(
        val width: Int,
        val height: Int
    )

    data class ViewportConfiguration(
        val viewportWidth: Int,
        val viewportHeight: Int,
        val contentWidth: Int,
        val contentHeight: Int,
        val scrollX: Int,
        val scrollY: Int
    ) {
        fun clamp(): ViewportConfiguration {
            val clampedViewportWidth = viewportWidth.coerceAtLeast(0)
            val clampedViewportHeight = viewportHeight.coerceAtLeast(0)
            val clampedContentWidth = contentWidth.coerceAtLeast(clampedViewportWidth)
            val clampedContentHeight = contentHeight.coerceAtLeast(clampedViewportHeight)
            val maxScrollX = (clampedContentWidth - clampedViewportWidth).coerceAtLeast(0)
            val maxScrollY = (clampedContentHeight - clampedViewportHeight).coerceAtLeast(0)
            return copy(
                viewportWidth = clampedViewportWidth,
                viewportHeight = clampedViewportHeight,
                contentWidth = clampedContentWidth,
                contentHeight = clampedContentHeight,
                scrollX = scrollX.coerceIn(0, maxScrollX),
                scrollY = scrollY.coerceIn(0, maxScrollY)
            )
        }
    }

    class NativeEmbeddedSurfaceControlHost(
        private val context: Context,
        private val pluginPackageName: String,
        private val pluginId: String,
        private val instanceId: Int
    ) : AutoCloseable {
        private val surfaceControlClient = AudioPluginHostHelper.createSurfaceControl(context)

        val surfaceView: View
            get() = surfaceControlClient.surfaceView

        @RequiresApi(Build.VERSION_CODES.R)
        suspend fun getPreferredSizeOrFallback(fallbackWidth: Int, fallbackHeight: Int): Size {
            val preferredSize = surfaceControlClient.getPreferredSizeNoHandler(
                pluginPackageName,
                pluginId,
                instanceId
            )
            val width = preferredSize.getOrNull(0)?.takeIf { it > 0 } ?: fallbackWidth
            val height = preferredSize.getOrNull(1)?.takeIf { it > 0 } ?: fallbackHeight
            return Size(width, height)
        }

        @RequiresApi(Build.VERSION_CODES.R)
        suspend fun connect(width: Int, height: Int, withIdleHandler: Boolean = false) {
            if (withIdleHandler)
                surfaceControlClient.connectUI(pluginPackageName, pluginId, instanceId, width, height)
            else
                surfaceControlClient.connectUINoHandler(pluginPackageName, pluginId, instanceId, width, height)
        }

        fun show() {
            surfaceView.visibility = View.VISIBLE
        }

        fun hide() {
            surfaceView.visibility = View.GONE
        }

        @RequiresApi(Build.VERSION_CODES.R)
        fun resize(width: Int, height: Int) {
            surfaceControlClient.resizeUI(instanceId, width, height)
        }

        @RequiresApi(Build.VERSION_CODES.R)
        fun configureViewport(configuration: ViewportConfiguration) {
            val clamped = configuration.clamp()
            surfaceControlClient.configureViewport(
                instanceId,
                clamped.viewportWidth,
                clamped.viewportHeight,
                clamped.contentWidth,
                clamped.contentHeight,
                clamped.scrollX,
                clamped.scrollY
            )
        }

        override fun close() {
            surfaceControlClient.close()
        }
    }
}
